WikiPDFExport: Cannot find the requested file xx to export

### Azure DevOps Extensions

WIKI PDF Generator

### Platform

Azure DevOps Server (TFS) - Specify Version

### Azure DevOps Server (TFS) Version

Azure DevOps Server 2022

### Extension Version

3.*

### Describe the bug

I did setup the task in a Release Pipeline and it works as long as o don't provide a Filename in the "Single file to export" option.
As soon as I put a filename in there, the error "Cannot find the requested file xx to export" is shown and it fails.

I double and tripple checked, if the file is in the artifacts folder of the release agent- it definitly is.

An odd behaviour also occurs if I export the whole Wiki to the pdf. All files but the "leaves" of the folder tree get generated into the pdf. I'm not sure, if this has anything to do with the my problem.

Example folder structure where Home.md and Versions.md will be included, but ReleaseNotes_Release-39.md not:
- File Home.md
- Folder Home
  - File Versions.md
  - Folder Versions
    - ReleaseNotes_Release-39.md

Configuration of the Task:
```
variables:
  ReleaseNotesPdfFileName: 'ReleaseNotes.pdf'

steps:
- task: richardfennellBM.BM-VSTS-WikiPDFExport-Tasks.WikiPDFExportTask.WikiPdfExportTask@3
  displayName: 'Create PDF from Wikipage'
  inputs:
    cloneRepo: false
    singleFile: 'ReleaseNotes_Release-95.md'
    useAgentToken: true
    localpath: '$(System.DefaultWorkingDirectory)\wikiRepo'
    rootExportPath: '$(System.DefaultWorkingDirectory)\wikiRepo\Home\Versions'
    outputFile: '$(System.DefaultWorkingDirectory)\$(ReleaseNotesPdfFileName)'
    injectExtraHeader: true
```

What am I missing?

### Logging Information

2023-12-04T10:44:02.1533564Z URL used https://xxx.xxx.xxx/DefaultCollection/SU/_git/SU.wiki
2023-12-04T10:44:02.2212486Z Cleaned C:\azagent\agent2\_work\r50\a\wikiRepo
2023-12-04T10:44:03.3620611Z Cloned xxx.xxx.xxx/DefaultCollection/SU/_git/SU.wiki to C:\azagent\agent2\_work\r50\a\wikiRepo
2023-12-04T10:44:03.3633928Z A filename 'ReleaseNotes_Release-94.md' in the folder 'C:\azagent\agent2\_work\r50\a\wikiRepo\Home\Versions' has been requested so only processing that file 
2023-12-04T10:44:03.3700566Z ##[error]Cannot find the requested file ReleaseNotes_Release-94.md to export
2023-12-04T10:44:03.3713474Z ##[debug]Processed: ##vso[task.issue type=error;]Cannot find the requested file ReleaseNotes_Release-94.md to export
2023-12-04T10:44:03.3714797Z ##[debug]task result: Failed
2023-12-04T10:44:03.3716187Z ##[error]Cannot find the requested file ReleaseNotes_Release-94.md to export
2023-12-04T10:44:03.3717362Z ##[debug]Processed: ##vso[task.issue type=error;]Cannot find the requested file ReleaseNotes_Release-94.md to export
2023-12-04T10:44:03.3719961Z ##[debug]Processed: ##vso[task.complete result=Failed;]Cannot find the requested file ReleaseNotes_Release-94.md to export
2023-12-04T10:44:03.4250591Z ##[section]Finishing: Create PDF from Wikipage

I tried it the way you wrote and I'm getting another error now:

Unhandled: ENOENT: no such file or directory, chdir 'C:\azagent\agent2_work\r50\a' -> 'C:\azagent\agent2_work\r50\a\wikiRepo\Home\Versions\ReleaseNotes_Release-101.md'

### Logging Information
2023-12-05T13:22:22.5104004Z Processing the contents of the folder 'C:\azagent\agent2\_work\r50\a\wikiRepo\Home\Versions\ReleaseNotes_Release-101.md' 
2023-12-05T13:22:22.5111493Z No filename specified and cannot find the .order file in the root of the wiki, the exported PDF will therefore be empty
2023-12-05T13:22:22.5112409Z Adding the verbose flag to increase logging
2023-12-05T13:22:22.5117125Z Changing folder to C:\azagent\agent2\_work\r50\a\wikiRepo\Home\Versions\ReleaseNotes_Release-101.md
2023-12-05T13:22:22.5553475Z ##[error]Unhandled: ENOENT: no such file or directory, chdir 'C:\azagent\agent2\_work\r50\a' -> 'C:\azagent\agent2\_work\r50\a\wikiRepo\Home\Versions\ReleaseNotes_Release-101.md'
2023-12-05T13:22:22.5582089Z ##[error]Error: ENOENT: no such file or directory, chdir 'C:\azagent\agent2\_work\r50\a' -> 'C:\azagent\agent2\_work\r50\a\wikiRepo\Home\Versions\ReleaseNotes_Release-101.md'
    at process.wrappedChdir [as chdir] (node:internal/bootstrap/switches/does_own_process_state:112:14)
    at C:\azagent\agent2\_work\_tasks\WikiPdfExportTask_48d40d27-32e1-460e-8ea6-0dfb48abdd6f\3.4.3\ExportFunctions.js:93:17
    at Generator.next (<anonymous>)
    at C:\azagent\agent2\_work\_tasks\WikiPdfExportTask_48d40d27-32e1-460e-8ea6-0dfb48abdd6f\3.4.3\ExportFunctions.js:8:71
    at new Promise (<anonymous>)
    at __awaiter (C:\azagent\agent2\_work\_tasks\WikiPdfExportTask_48d40d27-32e1-460e-8ea6-0dfb48abdd6f\3.4.3\ExportFunctions.js:4:12)
    at ExportPDF (C:\azagent\agent2\_work\_tasks\WikiPdfExportTask_48d40d27-32e1-460e-8ea6-0dfb48abdd6f\3.4.3\ExportFunctions.js:39:12)
    at C:\azagent\agent2\_work\_tasks\WikiPdfExportTask_48d40d27-32e1-460e-8ea6-0dfb48abdd6f\3.4.3\ExportFunctions.js:168:13
    at Generator.next (<anonymous>)
    at C:\azagent\agent2\_work\_tasks\WikiPdfExportTask_48d40d27-32e1-460e-8ea6-0dfb48abdd6f\3.4.3\ExportFunctions.js:8:71
2023-12-05T13:22:22.5650902Z ##[section]Finishing: Create PDF from WIKI

While I'm still not able to get the PDF file from a single MD file, at least it gave me a clue, why all but the "leaf documents" are missing in the generated pdf --> the .order file is missing in the folder "Versions". I know this is another issue, but who's responsibility is it to create and update that file? Should it be done automatically by the Azure DevOps Wiki on adding a new file or is this a manual task?

--- a/Extensions/WikiPDFExport/WikiPDFExportTask/src/ExportFunctions.ts
+++ b/Extensions/WikiPDFExport/WikiPDFExportTask/src/ExportFunctions.ts
@@ -72,20 +72,28 @@ export async function ExportPDF(
                 logError(`Cannot find wiki folder ${wikiRootPath}`);
                 return;
             } else {
+                logInfo(`Using ${wikiRootPath} as the root path to export`);
                 args += ` -p "${wikiRootPath}"`;
             }
         }
 
         if (singleFile.length > 0) {
-            if (!fs.existsSync(`${singleFile}`)) {
+            // first check for a fully specified path
+            if (fs.existsSync(`${singleFile}`)) {
+                logInfo(`Found file using the fully qualified path of ${singleFile}`);
+                args += ` -s "${singleFile}"`;
+            } else if (fs.existsSync(`${path.join(wikiRootPath, singleFile)}`)) {
+                logInfo(`Found file using the expanded relative path of ${path.join(wikiRootPath, singleFile)}`);
+                args += ` -s "${singleFile}"`; // we don't to add the wikiRootPath as it is already included in the -p parameter
+            } else {
                 logError(`Cannot find the requested file ${singleFile} to export`);
                 return;
-            } else {
-                args += ` -s "${singleFile}"`;
             }
         } else {
-            if (!fs.existsSync(`${wikiRootPath}/.order`)) {
-                logInfo(`No filename specified and cannot find the .order file in the root of the wiki, the exported PDF will therefore be empty`);
+            if (!fs.existsSync(`${path.join(wikiRootPath, ".order")}`)) {
+                logWarning(`No filename specified and cannot find the .order file in the root of the wiki, the exported PDF will therefore be empty. Consider setting the underlying AzureDevOps.WikiPDFExport tools property '--include-unlisted-pages' using the 'ExtraParameters' parameter`);
+            } else {
+                logInfo(`Using the .order file in the root of the wiki to determine the order of the pages to export`);
             }
         }
 

--- a/Extensions/WikiPDFExport/WikiPDFExportTask/src/ExportFunctions.ts
+++ b/Extensions/WikiPDFExport/WikiPDFExportTask/src/ExportFunctions.ts
@@ -80,7 +80,7 @@ export async function ExportPDF(
         if (singleFile.length > 0) {
             // first check for a fully specified path
             if (fs.existsSync(`${singleFile}`)) {
-                logInfo(`Found file using the fully qualified path of ${singleFile}`);
+                logInfo(`Found file the file ${singleFile} relative to the current working directory`);
                 args += ` -s "${singleFile}"`;
             } else if (fs.existsSync(`${path.join(wikiRootPath, singleFile)}`)) {
                 logInfo(`Found file using the expanded relative path of ${path.join(wikiRootPath, singleFile)}`);

--- a/Extensions/WikiPDFExport/WikiPDFExportTask/task/task.json
+++ b/Extensions/WikiPDFExport/WikiPDFExportTask/task/task.json
@@ -62,7 +62,7 @@
       "label": "Single file to export",
       "defaultValue": "",
       "required": false,
-      "helpMarkDown": "Name of the single file to be exported, default is empty i.e. whole repo is exported e.g. myfile.md"
+      "helpMarkDown": "Single file to export. Can be set as either a relative path to a file in the `localPath` or `RootExportPath` folder e.g. page.md or as fully specified path. If not set the whole WIKI is exported based on the structure details the `.order` file in the root of the repo",
     },
     {
       "name": "useAgentToken",
@@ -105,7 +105,7 @@
       "label": "Local folder",
       "defaultValue": "",
       "required": false,
-      "helpMarkDown": "The path to the root of the cloned the repo if exporting the whole repo, a folder within the repo to export part of the repo or finally the folder containing a single file to export. For this final option the filename must be specified below"
+      "helpMarkDown": "If not set, this defaults to the path of the root of the cloned the repo. It can be set to a folder within the cloned repo export only part of the repo. If only a single file is required the a filename can be specified using the singleFile parameter"
     },
     {
       "name": "outputFile",

--- a/Extensions/WikiPDFExport/WikiPDFExportTask/task/task.json
+++ b/Extensions/WikiPDFExport/WikiPDFExportTask/task/task.json
@@ -62,7 +62,7 @@
       "label": "Single file to export",
       "defaultValue": "",
       "required": false,
-      "helpMarkDown": "Single file to export. Can be set as either a relative path to a file in the `localPath` or `RootExportPath` folder e.g. page.md or as fully specified path. If not set the whole WIKI is exported based on the structure details the `.order` file in the root of the repo",
+      "helpMarkDown": "Single file to export. Can be set as either a relative path to a file in the `localPath` or `RootExportPath` folder e.g. page.md or as fully specified path. If not set the whole WIKI is exported based on the structure details the `.order` file in the root of the repo"
     },
     {
       "name": "useAgentToken",

--- a/Extensions/WikiPDFExport/readme.md
+++ b/Extensions/WikiPDFExport/readme.md
@@ -7,7 +7,7 @@ When run, the task will download the current release (or optionally a pre-releas
    - Optionally clone a Git based WIKI repo that is hosted on Azure DevOps or GitHub into a local folder for exporting
    - Or if you already have the folder structure on the agent then this feature can be skipped
 2. The task will then export either
-   - the whole WIKI structure as a PDF (based on the .order file is present in the root)
+   - the whole WIKI structure as a PDF (based on the .order file which must be present in the repo root)
    - a single named file
 
 ## How the AzureDevOps.WikiPDFExport tool is obtained
@@ -60,8 +60,8 @@ steps:
 
 ## Parameters
 ### AzureDevOps.WikiPDFExport Specific
-- SingleFile - Optional single file to export in the localPath folder e.g. page.md
-- RootExportPath - The path to the root of the cloned the repo if exporting the whole repo, a folder within the repo to export part of the repo or finally the folder containing a single file to export. For this final option the filename must be specified below
+- SingleFile - Single file to export. Can be set as either a relative path to a file in the `localPath` or `RootExportPath` folder e.g. page.md or as fully specified path. If not set the whole WIKI is exported based on the structure details the `.order` file in the root of the repo.
+- RootExportPath - Optional: If not set, this defaults to the path of the root of the cloned the repo. It can be set to a folder within the cloned repo export only part of the repo. If only a single file is required the a filename can be specified using the `singleFile`` parameter
 - ExtraParameters - Any optional extra as defined at [WikiPDFExport](https://github.com/MaxMelcher/AzureDevOps.WikiPDFExport/) you wish to pass to the command line tool - noting that this task automatically manages the -p, -s, -c and -v parameters
 - usePreRelease - If set to true pre-release version of the [AzureDevOps.WikiPDFExport tool](https://github.com/MaxMelcher/AzureDevOps.WikiPDFExport) tool will be used
 - overrideExePath - An optional path to a previously download copy of the [AzureDevOps.WikiPDFExport tool](https://github.com/MaxMelcher/AzureDevOps.WikiPDFExport). If not set the task will download the current release of this tool

--- a/Extensions/WikiPDFExport/test-stage.yml
+++ b/Extensions/WikiPDFExport/test-stage.yml
@@ -50,6 +50,15 @@ steps:
     outputFile: '$(Build.ArtifactStagingDirectory)/PDF/$(Agent.OS)-singleFile.pdf'
     downloadPath: '${{parameters.localpath}}'
 - task: richardfennellBM.BM-VSTS-WikiPDFExport-Tasks-DEV.WikiPDFExportTask.WikiPdfExportTask@3
+  displayName: 'Export Single File'
+  inputs:
+    cloneRepo: false
+    usePreRelease: false
+    localpath: '$(System.DefaultWorkingDirectory)'
+    singleFile: 'inline.md'
+    outputFile: '$(Build.ArtifactStagingDirectory)/PDF/$(Agent.OS)-singleFile.pdf'
+    downloadPath: '${{parameters.localpath}}'
+- task: richardfennellBM.BM-VSTS-WikiPDFExport-Tasks-DEV.WikiPDFExportTask.WikiPdfExportTask@3
   displayName: 'Export Public GitHub WIKI'
   condition: succeededOrFailed()
   inputs:
@@ -79,6 +88,18 @@ steps:
     localpath: '$(System.DefaultWorkingDirectory)/repopartial'
     rootExportPath: '$(System.DefaultWorkingDirectory)/repopartial/folder'
     downloadPath: '${{parameters.localpath}}'
+- task: richardfennellBM.BM-VSTS-WikiPDFExport-Tasks-DEV.WikiPDFExportTask.WikiPdfExportTask@3
+  displayName: 'Export single part of the Azure DevOps WIKI'
+  condition: succeededOrFailed()
+  inputs:
+    cloneRepo: true
+    repo: 'https://dev.azure.com/richardfennell/GitHub/_git/GitHub.wiki'
+    useAgentToken: true
+    outputFile: '$(Build.ArtifactStagingDirectory)/PDF/$(Agent.OS)-AzrepoFilterSingle.pdf'
+    localpath: '$(System.DefaultWorkingDirectory)/repopartial1'
+    rootExportPath: '$(System.DefaultWorkingDirectory)/repopartial1/folder'
+    downloadPath: '${{parameters.localpath}}'
+    singlefile: 'test1.md'
 - task: richardfennellBM.BM-VSTS-WikiPDFExport-Tasks-DEV.WikiPDFExportTask.WikiPdfExportTask@3
   displayName: 'Export Azure DevOps WIKI without downloading tool'
   condition: and(succeededOrFailed(), eq(variables['AGENT.OS'], 'Windows_NT'))

